From 1c071e4006f725b4a2b04ddcfb1891320912e95d Mon Sep 17 00:00:00 2001
From: Xaver Hugl <xaver.hugl@gmail.com>
Date: Mon, 19 May 2025 14:25:56 +0200
Subject: [PATCH] core/iccprofile: also estimate black point even if there's no
 luminance tag

We can still get a relative value out of it

(cherry picked from commit 0ee6eb5fb3b4389f187d8b9784d5e23b4c883f84)
---
 src/backends/drm/drm_output.cpp |  2 +-
 src/core/iccprofile.cpp         | 20 ++++++++++----------
 src/core/iccprofile.h           |  6 +++---
 3 files changed, 14 insertions(+), 14 deletions(-)

diff --git a/src/backends/drm/drm_output.cpp b/src/backends/drm/drm_output.cpp
index 151cb2d466f..f2024b223db 100644
--- a/src/backends/drm/drm_output.cpp
+++ b/src/backends/drm/drm_output.cpp
@@ -427,8 +427,8 @@ std::pair<ColorDescription, QVector3D> DrmOutput::createColorDescription(const s
     const double sdrGamutWideness = props->sdrGamutWideness.value_or(m_state.sdrGamutWideness);
     const auto iccProfile = props->iccProfile.value_or(m_state.iccProfile);
     if (colorSource == ColorProfileSource::ICC && !effectiveHdr && !effectiveWcg && iccProfile) {
-        const double minBrightness = iccProfile->minBrightness().value_or(0);
         const double maxBrightness = iccProfile->maxBrightness().value_or(200);
+        const double minBrightness = iccProfile->relativeBlackPoint().value_or(0) * maxBrightness;
         const auto sdrColor = Colorimetry::fromName(NamedColorimetry::BT709).interpolateGamutTo(iccProfile->colorimetry(), sdrGamutWideness);
         const bool allowSdrSoftwareBrightness = props->allowSdrSoftwareBrightness.value_or(m_state.allowSdrSoftwareBrightness);
         const double brightnessFactor = (!m_brightnessDevice && allowSdrSoftwareBrightness) ? brightness : 1.0;
diff --git a/src/core/iccprofile.cpp b/src/core/iccprofile.cpp
index 91ab1f92c64..17be36c93d5 100644
--- a/src/core/iccprofile.cpp
+++ b/src/core/iccprofile.cpp
@@ -27,14 +27,14 @@ static const Colorimetry CIEXYZD50 = Colorimetry{
 
 const ColorDescription IccProfile::s_connectionSpace = ColorDescription(CIEXYZD50, TransferFunction(TransferFunction::linear, 0, 1), 1, 0, 1, 1);
 
-IccProfile::IccProfile(cmsHPROFILE handle, const Colorimetry &colorimetry, std::optional<ColorPipeline> &&bToA0Tag, std::optional<ColorPipeline> &&bToA1Tag, const std::shared_ptr<ColorTransformation> &inverseEOTF, const std::shared_ptr<ColorTransformation> &vcgt, std::optional<double> minBrightness, std::optional<double> maxBrightness)
+IccProfile::IccProfile(cmsHPROFILE handle, const Colorimetry &colorimetry, std::optional<ColorPipeline> &&bToA0Tag, std::optional<ColorPipeline> &&bToA1Tag, const std::shared_ptr<ColorTransformation> &inverseEOTF, const std::shared_ptr<ColorTransformation> &vcgt, std::optional<double> relativeBlackPoint, std::optional<double> maxBrightness)
     : m_handle(handle)
     , m_colorimetry(colorimetry)
     , m_bToA0Tag(std::move(bToA0Tag))
     , m_bToA1Tag(std::move(bToA1Tag))
     , m_inverseEOTF(inverseEOTF)
     , m_vcgt(vcgt)
-    , m_minBrightness(minBrightness)
+    , m_relativeBlackPoint(relativeBlackPoint)
     , m_maxBrightness(maxBrightness)
 {
 }
@@ -44,9 +44,9 @@ IccProfile::~IccProfile()
     cmsCloseProfile(m_handle);
 }
 
-std::optional<double> IccProfile::minBrightness() const
+std::optional<double> IccProfile::relativeBlackPoint() const
 {
-    return m_minBrightness;
+    return m_relativeBlackPoint;
 }
 
 std::optional<double> IccProfile::maxBrightness() const
@@ -341,16 +341,16 @@ IccProfile::Expected IccProfile::load(const QString &path)
         return Expected(i18n("ICC profile \"%1\" is broken, its primaries are invalid", path));
     }
 
-    std::optional<double> minBrightness;
     std::optional<double> maxBrightness;
     if (cmsCIEXYZ *luminance = static_cast<cmsCIEXYZ *>(cmsReadTag(handle, cmsSigLuminanceTag))) {
         // for some reason, lcms exposes the luminance as a XYZ triple...
         // only Y is non-zero, and it's the brightness in nits
         maxBrightness = luminance->Y;
-        cmsCIEXYZ blackPoint;
-        if (cmsDetectDestinationBlackPoint(&blackPoint, handle, INTENT_RELATIVE_COLORIMETRIC, 0)) {
-            minBrightness = blackPoint.Y * luminance->Y;
-        }
+    }
+    std::optional<double> relativeBlackPoint;
+    cmsCIEXYZ blackPoint;
+    if (cmsDetectDestinationBlackPoint(&blackPoint, handle, INTENT_RELATIVE_COLORIMETRIC, 0)) {
+        relativeBlackPoint = blackPoint.Y;
     }
 
     if (cmsIsTag(handle, cmsSigBToD1Tag) && !cmsIsTag(handle, cmsSigBToA1Tag) && !cmsIsTag(handle, cmsSigBToA0Tag)) {
@@ -404,7 +404,7 @@ IccProfile::Expected IccProfile::load(const QString &path)
     std::vector<std::unique_ptr<ColorPipelineStage>> stages;
     stages.push_back(std::make_unique<ColorPipelineStage>(cmsStageAllocToneCurves(nullptr, toneCurves.size(), toneCurves.data())));
     const auto inverseEOTF = std::make_shared<ColorTransformation>(std::move(stages));
-    return std::make_unique<IccProfile>(handle, Colorimetry(red, green, blue, white), std::move(bToA0), std::move(bToA1), inverseEOTF, vcgt, minBrightness, maxBrightness);
+    return std::make_unique<IccProfile>(handle, Colorimetry(red, green, blue, white), std::move(bToA0), std::move(bToA1), inverseEOTF, vcgt, relativeBlackPoint, maxBrightness);
 }
 
 }
diff --git a/src/core/iccprofile.h b/src/core/iccprofile.h
index be4e2e4a74c..bc4057acfd9 100644
--- a/src/core/iccprofile.h
+++ b/src/core/iccprofile.h
@@ -25,7 +25,7 @@ class ColorLUT3D;
 class KWIN_EXPORT IccProfile
 {
 public:
-    explicit IccProfile(cmsHPROFILE handle, const Colorimetry &colorimetry, std::optional<ColorPipeline> &&bToA0Tag, std::optional<ColorPipeline> &&bToA1Tag, const std::shared_ptr<ColorTransformation> &inverseEOTF, const std::shared_ptr<ColorTransformation> &vcgt, std::optional<double> minBrightness, std::optional<double> maxBrightness);
+    explicit IccProfile(cmsHPROFILE handle, const Colorimetry &colorimetry, std::optional<ColorPipeline> &&bToA0Tag, std::optional<ColorPipeline> &&bToA1Tag, const std::shared_ptr<ColorTransformation> &inverseEOTF, const std::shared_ptr<ColorTransformation> &vcgt, std::optional<double> relativeBlackPoint, std::optional<double> maxBrightness);
     ~IccProfile();
 
     /**
@@ -44,7 +44,7 @@ public:
      */
     std::shared_ptr<ColorTransformation> vcgt() const;
     const Colorimetry &colorimetry() const;
-    std::optional<double> minBrightness() const;
+    std::optional<double> relativeBlackPoint() const;
     std::optional<double> maxBrightness() const;
 
     // TODO Plasma 6.4 port this back to std::expected
@@ -72,7 +72,7 @@ private:
     const std::optional<ColorPipeline> m_bToA1Tag;
     const std::shared_ptr<ColorTransformation> m_inverseEOTF;
     const std::shared_ptr<ColorTransformation> m_vcgt;
-    const std::optional<double> m_minBrightness;
+    const std::optional<double> m_relativeBlackPoint;
     const std::optional<double> m_maxBrightness;
 };
 
-- 
GitLab

